File: javascript/src/LazyCollectionHolderOf1.ts
//··························································
// Copyright (c) 2023-2026. Jonathan Bédard ~ JóôòKiwi
//
// This project is free to use.
// All the right is reserved to the author of this project.
// My projects:
//  - https://github.com/joooKiwi/type
//  - https://github.com/joooKiwi/lazy
//  - https://github.com/joooKiwi/collection
//  - https://github.com/joooKiwi/enumeration
//··························································

import type {CollectionHolder}    from "./CollectionHolder"
import type {CollectionHolderOf1} from "./CollectionHolderOf1"
import type {Optional}            from "./optional/Optional"

import {AbstractCollectionHolderOf1} from "./AbstractCollectionHolderOf1"
import {LazyCollectionHolderOf0Or1}  from "./LazyCollectionHolderOf0Or1"

const FAIL_CALLBACK: () => never = () => { throw new ReferenceError("This callback is never supposed to be called normally.",) }

/**
 * An instance of [CollectionHolder] with only a single value but lazily retrieved.
 *
 * It does retrieve the value only once from the callback received in its `constructor`.
 *
 * @typeParam T The type (by default `unknown`)
 * @see SingleValueCollectionHolder
 * @see ArrayOf1AsCollectionHolder
 */
export class LazyCollectionHolderOf1<const T = unknown, >
    extends AbstractCollectionHolderOf1<T> {

    //#region -------------------- Field --------------------

    #lateValue: () => T
    #value?: T
    #isValueInitialized: boolean
    #hasNull?: boolean
    #hasNoNulls?: boolean

    //#endregion -------------------- Field --------------------
    //#region -------------------- Constructor --------------------

    public constructor(lateValue: () => T,) {
        super()
        this.#lateValue = lateValue
        this.#isValueInitialized = false
    }

    //#endregion -------------------- Constructor --------------------
    //#region -------------------- Methods --------------------

    public override _create<const U, >(lateValue: () => U,): CollectionHolderOf1<U> {
        return new LazyCollectionHolderOf1(lateValue,)
    }

    protected override _create0Or1<const U, >(latePossibleValue: () => Optional<U>,): CollectionHolder<U> {
        return new LazyCollectionHolderOf0Or1(latePossibleValue,)
    }

    /** The internal value passed through the {@link constructor} */
    public override get 0() { return this.value }

    /** The internal value passed through the {@link constructor} */
    public override get value(): T {
        if (this.#isValueInitialized)
            return this.#value as T

        const value = this.#value = this.#lateValue()
        this.#lateValue = FAIL_CALLBACK // We do not need the callback anymore once the value has been retrieved
        this.#isValueInitialized = true
        return value
    }

    public override get hasNull(): boolean { return this.#hasNull ?? super.hasNull }

    public override get hasNoNulls(): boolean { return this.#hasNoNulls ?? super.hasNoNulls }

    //#endregion -------------------- Methods --------------------

}

File: java/collection.joookiwi/src/main/java/joookiwi/collection/java/LazyCollectionHolderOf1.java
package joookiwi.collection.java;

import java.util.function.Consumer;
import java.util.function.IntConsumer;
import java.util.function.ObjIntConsumer;
import java.util.function.Supplier;
import joookiwi.collection.java.callback.IntObjConsumer;
import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.MustBeInvokedByOverriders;
import org.jetbrains.annotations.NotNullByDefault;
import org.jetbrains.annotations.Nullable;

import static joookiwi.collection.java.CommonContracts.ALWAYS_NEW_0;
import static joookiwi.collection.java.CommonContracts.ALWAYS_THIS_1;

/// An instance of [CollectionHolder] with only a single value but lazily retrieved.
///
/// It does retrieve the value only once from the [Supplier] received in its `constructor`.
///
/// @param <T> The type
/// @see CollectionHolderOf1
/// @see ArrayOf1AsCollectionHolder
@NotNullByDefault
public class LazyCollectionHolderOf1<T extends @Nullable Object>
        extends AbstractCollectionHolderOf1<T> {

    //#region -------------------- Fields --------------------

    private static final Supplier<?> FAIL_CALLBACK = () -> { throw new RuntimeException("This callback is never supposed to be called normally."); };

    private Supplier<? extends T> __lateValue;
    private boolean __isValueInitialized;
    private @Nullable T __value;

    private boolean __hasNull;
    private boolean __hasNullInitialized;

    private boolean __hasNoNulls;
    private boolean __hasNoNullsInitialized;

    //#endregion -------------------- Fields --------------------
    //#region -------------------- Constructor --------------------

    public LazyCollectionHolderOf1(final Supplier<? extends T> lateValue) {
        super();
        __lateValue = lateValue;
        __isValueInitialized = false;
    }

    //#endregion -------------------- Constructor --------------------
    //#region -------------------- Methods --------------------

    /// The internal value passed through the `constructor`
    @SuppressWarnings("unchecked cast")
    @Override public T value() {
        if (__isValueInitialized)
            return __value;

        final var value = __value = __lateValue.get();
        __lateValue = (Supplier<? extends T>) FAIL_CALLBACK; // We do not need the callback anymore once the value has been retrieved
        __isValueInitialized = true;
        return value;
    }

    @Override public boolean hasNull() {
        if (__hasNullInitialized)
            return __hasNull;

        final var value = __hasNull = value() == null;
        __hasNullInitialized = true;
        return value;
    }

    @Override public boolean hasNoNulls() {
        if (__hasNoNullsInitialized)
            return __hasNoNulls;

        final var value = __hasNoNulls = value() != null;
        __hasNoNullsInitialized = true;
        return value;
    }

    @Contract(ALWAYS_THIS_1) @Override public LazyCollectionHolderOf1<T> onEach(final ObjIntConsumer<? super T> action) { return (LazyCollectionHolderOf1<T>) super.onEach(action); }
    @Contract(ALWAYS_THIS_1) @Override public LazyCollectionHolderOf1<T> onEach(final Consumer<? super T>       action) { return (LazyCollectionHolderOf1<T>) super.onEach(action); }
    @Contract(ALWAYS_THIS_1) @Override public LazyCollectionHolderOf1<T> onEach(final Runnable                  action) { return (LazyCollectionHolderOf1<T>) super.onEach(action); }

    @Contract(ALWAYS_THIS_1) @Override public LazyCollectionHolderOf1<T> onEachIndexed(final IntObjConsumer<? super T> action) { return (LazyCollectionHolderOf1<T>) super.onEachIndexed(action); }
    @Contract(ALWAYS_THIS_1) @Override public LazyCollectionHolderOf1<T> onEachIndexed(final IntConsumer               action) { return (LazyCollectionHolderOf1<T>) super.onEachIndexed(action); }
    @Contract(ALWAYS_THIS_1) @Override public LazyCollectionHolderOf1<T> onEachIndexed(final Runnable                  action) { return (LazyCollectionHolderOf1<T>) super.onEachIndexed(action); }



    @MustBeInvokedByOverriders
    @Contract(ALWAYS_NEW_0)
    @Override public LazyCollectionHolderOf1<T> clone() { return (LazyCollectionHolderOf1<T>) super.clone(); }

//    @MustBeInvokedByOverriders
//    @Contract(ALWAYS_NEW_0)
//    @Override public LazyCollectionHolderOf1<T> shallowClone() { return clone(); }
//
//
//    @MustBeInvokedByOverriders
//    @Contract(ALWAYS_NEW_0)
//    @Override public LazyCollectionHolderOf1<T> deepClone() { return new LazyCollectionHolderOf1<>(() -> CloneInvokator.getInstance().cloneIfCloneable(value())); }

    //#endregion -------------------- Methods --------------------

}
